### PR TITLE
Ftr koski integration

### PR DESCRIPTION
**Pull request by anttiviljakainen**

### CODE REVIEW

File: framework/src/main/java/fi/otavanopisto/pyramus/koski/KoskiStudentHandler.java [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+        return ArviointiasteikkoYleissivistava.GRADE_S;
+    }
+    
+    return null;
+  }
+
+  protected boolean isPakollinenOppiaine(Student student, KoskiOppiaineetYleissivistava oppiaine) {
+    Set<KoskiOppiaineetYleissivistava> pakollisetOppiaineet = settings.getPakollisetOppiaineet(student.getStudyProgramme().getId());
+    if (pakollisetOppiaineet != null)
+      return pakollisetOppiaineet.contains(oppiaine);
```
Comment (ilmoeuro) [inline framework/src/main/java/fi/otavanopisto/pyramus/koski/KoskiStudentHandler.java]: Use braces {} - see https://www.dwheeler.com/essays/apple-goto-fail.html

File: framework/src/main/java/fi/otavanopisto/pyramus/koski/KoskiStudentHandler.java [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+    // TODO Vapautetut aineet
```
Comment (ilmoeuro) [inline framework/src/main/java/fi/otavanopisto/pyramus/koski/KoskiStudentHandler.java]: TODO ?

File: framework/src/main/java/fi/otavanopisto/pyramus/koski/KoskiStudentHandler.java [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+    String lang = userVariableDAO.findByUserAndKey(student, "lukioAidinkieli");
+    String aLang = userVariableDAO.findByUserAndKey(student, "lukioKieliA");
+    String a1Lang = userVariableDAO.findByUserAndKey(student, "lukioKieliA1");
+    String a2Lang = userVariableDAO.findByUserAndKey(student, "lukioKieliA2");
+    String b1Lang = userVariableDAO.findByUserAndKey(student, "lukioKieliB1");
+    String b2Lang = userVariableDAO.findByUserAndKey(student, "lukioKieliB2");
+    String b3Lang = userVariableDAO.findByUserAndKey(student, "lukioKieliB3");
+    String religion = userVariableDAO.findByUserAndKey(student, "lukioUskonto");
+
+    if (StringUtils.isNotBlank(math))
```
Comment (ilmoeuro) [inline framework/src/main/java/fi/otavanopisto/pyramus/koski/KoskiStudentHandler.java]: Use braces in this code block

File: framework/src/main/java/fi/otavanopisto/pyramus/koski/KoskiStudentHandler.java [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+      Course course = ca.getCourseStudent() != null ? ca.getCourseStudent().getCourse() : null;
+      Subject subject = course != null ? course.getSubject() : null;
+      
+      String courseCode = subject.getCode() + course.getCourseNumber();
+      CreditStub stub;
+      if (!stubs.containsKey(courseCode)) {
+        stub = new CreditStub(courseCode, course.getName(), subject);
+        stubs.put(courseCode, stub);
+      } else
+        stub = stubs.get(courseCode);
```
Comment (ilmoeuro) [inline framework/src/main/java/fi/otavanopisto/pyramus/koski/KoskiStudentHandler.java]: Use braces

File: framework/src/main/java/fi/otavanopisto/pyramus/koski/KoskiStudentHandler.java [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+    }
+    
+    for (TransferCredit tc : transferCredits) {
+      String courseCode = tc.getSubject().getCode() + tc.getCourseNumber();
+      CreditStub stub;
+      if (!stubs.containsKey(courseCode)) {
+        stub = new CreditStub(courseCode, tc.getCourseName(), tc.getSubject());
+        stubs.put(courseCode, stub);
+      } else
+        stub = stubs.get(courseCode);
```
Comment (ilmoeuro) [inline framework/src/main/java/fi/otavanopisto/pyramus/koski/KoskiStudentHandler.java]: Use braces

File: framework/src/main/java/fi/otavanopisto/pyramus/koski/KoskiStudentHandler.java [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+      }
+      
+      if (allNotNull(subject, courseNumber, courseName)) {
+        String courseCode = subject.getCode() + courseNumber;
+        CreditStub stub;
+        if (!stubs.containsKey(courseCode)) {
+          stub = new CreditStub(courseCode, courseName, subject);
+          stubs.put(courseCode, stub);
+        } else
+          stub = stubs.get(courseCode);
```
Comment (ilmoeuro) [inline framework/src/main/java/fi/otavanopisto/pyramus/koski/KoskiStudentHandler.java]: Use braces

File: framework/src/main/java/fi/otavanopisto/pyramus/koski/KoskiStudentHandler.java [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+        if (ArviointiasteikkoYleissivistava.isNumeric(grade)) {
+          gradeSum += Integer.valueOf(grade.toString());
+          gradeCount++;
+        }
+      }
+      
+      if (gradeCount > 0) {
+        return ArviointiasteikkoYleissivistava.get(String.valueOf(Math.round((double) gradeSum / gradeCount)));
+      } else
+        return null;
```
Comment (ilmoeuro) [inline framework/src/main/java/fi/otavanopisto/pyramus/koski/KoskiStudentHandler.java]: Use braces - especially important here with two "else" clauses after each other

File: framework/src/main/java/fi/otavanopisto/pyramus/koski/KoskiStudentHandler.java [hunk 8]
```diff
@@ -0,0 +1,315 @@
...
+    } else {
+      if (grades.stream().anyMatch(grade -> ArviointiasteikkoYleissivistava.isLiteral(grade))) {
+        // Literal grade S/H
+        
+        if (grades.stream().anyMatch(grade -> ArviointiasteikkoYleissivistava.GRADE_S == grade))
+          return ArviointiasteikkoYleissivistava.GRADE_S;
```
Comment (ilmoeuro) [inline framework/src/main/java/fi/otavanopisto/pyramus/koski/KoskiStudentHandler.java]: Use braces

File: framework/src/main/java/fi/otavanopisto/pyramus/koski/KoskiStudentHandler.java [hunk 9]
```diff
@@ -0,0 +1,315 @@
...
+        else
+          return ArviointiasteikkoYleissivistava.GRADE_H;
+      } else
+        return null;
```
Comment (ilmoeuro) [inline framework/src/main/java/fi/otavanopisto/pyramus/koski/KoskiStudentHandler.java]: Use braces

File: framework/src/main/java/fi/otavanopisto/pyramus/koski/koodisto/ArviointiasteikkoYleissivistava.java [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+    return ArrayUtils.contains(LITERAL, arvosana);
+  }
+  
+  public static ArviointiasteikkoYleissivistava bestGrade(Collection<ArviointiasteikkoYleissivistava> collection) {
+    ArviointiasteikkoYleissivistavaComparator comparator = new ArviointiasteikkoYleissivistavaComparator();
+    ArviointiasteikkoYleissivistava bestGrade = null;
+    
+    for (ArviointiasteikkoYleissivistava grade : collection) {
+      if (bestGrade == null || comparator.compare(grade, bestGrade) > 0)
+        bestGrade = grade;
```
Comment (ilmoeuro) [inline framework/src/main/java/fi/otavanopisto/pyramus/koski/koodisto/ArviointiasteikkoYleissivistava.java]: Braces

File: framework/src/main/java/fi/otavanopisto/pyramus/koski/model/lukio/LukionOpiskeluoikeudenLisatiedot.java [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+  @JsonProperty("sisäoppilaitosmainenMajoitus")
+  public List<Majoitusjakso> getSisaoppilaitosmainenMajoitus() {
+    return sisaoppilaitosmainenMajoitus;
+  }
+
+  private final boolean pidennettyPaattymispaiva;
+  private final boolean ulkomainenVaihtoopiskelija;
+  private Kuvaus alle18vuotiaanAikuistenLukiokoulutuksenAloittamisenSyy;
+  private final boolean yksityisopiskelija;
+//  private List<ErityisenKoulutustehtavanJaksot> erityisenKoulutustehtävänJaksot;
```
Comment (ilmoeuro) [inline framework/src/main/java/fi/otavanopisto/pyramus/koski/model/lukio/LukionOpiskeluoikeudenLisatiedot.java]: Commented out code?

File: rest/src/main/java/fi/otavanopisto/pyramus/rest/filter/SecurityFilter.java [hunk 12]
```diff
@@ -60,6 +60,8 @@ public void filter(ContainerRequestContext requestContext) throws IOException {
     if (method == null){
       requestContext.abortWith(Response.status(javax.ws.rs.core.Response.Status.INTERNAL_SERVER_ERROR).build());
     } else {
+//      logger.log(Level.INFO, "REST: " + requestContext.getUriInfo().getPath());
```
Comment (ilmoeuro) [inline rest/src/main/java/fi/otavanopisto/pyramus/rest/filter/SecurityFilter.java]: Commented out code?

File: framework/src/main/java/fi/otavanopisto/pyramus/koski/koodisto/OppiaineAidinkieliJaKirjallisuus.java [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+  AI4,
+  AI5,
+  AI6,
+  AI7,
+  AI8,
+  AI9,
+  AI10,
+  AI11,
+  AI12,
+  AIAI
```
[the remark this answers is not in the document]
Comment (anttiviljakainen) [reply framework/src/main/java/fi/otavanopisto/pyramus/koski/koodisto/OppiaineAidinkieliJaKirjallisuus.java]: https://extra.koski.opintopolku.fi/koski/documentation/koodisto/oppiaineaidinkielijakirjallisuus/latest

[the remark this answers is not in the document]
Comment (pkukkon) [reply framework/src/main/java/fi/otavanopisto/pyramus/koski/koodisto/OppiaineAidinkieliJaKirjallisuus.java]: AI'll be damned...

**Pull request merged**